src/featbin/compute-filter.cc

```diff
@@ -0,0 +1,223 @@
+// featbin/compute-filters.cc
+
+// Copyright 2016 Pegah Ghahremani
+
+
+#include "feat/feature-functions.h"
+#include "matrix/srfft.h"
+#include "matrix/kaldi-matrix-inl.h"    
+#include "feat/wave-reader.h"
+namespace kaldi {
+
+// This function computes and accumulates  
+// dot-product of zero-mean input with its shifted version.
+// inner_prod[j] - inner product of input vector with its shifted version by j
+//                 sum_{i=0}^corr_window_size x[i] * x[i+j]
+// norms[j]      - inner product of shifted input by itself as sum_{i=0}^corr_window_size x[i+j]^2
+// lpc_order is the size of autocorr_coeffs.
+// If mean == 0, the wave is normalized using its mean, otherwise
+// it is mean-normalized using mean value.
+void AccStatsForCorrelation(const VectorBase<BaseFloat> &wave,
+                                int32 lpc_order, int32 corr_window_size,
+                                Vector<BaseFloat> *inner_prod,
+                                Vector<BaseFloat> *norms,
+                                BaseFloat mean) {
+  if (inner_prod->Dim() != lpc_order)
+    inner_prod->Resize(lpc_order);
+
+  if (norms->Dim() != lpc_order) 
+    norms->Resize(lpc_order);
+
+  Vector<BaseFloat> zero_mean_wave(wave);
+  if (mean != 0.0) 
+    zero_mean_wave.Add(mean);
+   else 
+    // subtract mean to normalize wave.
+    zero_mean_wave.Add(-wave.Sum() / wave.Dim());
+
+  // append zero et end if the length of shifted version 
+  // is less that corr_window_size.
+  zero_mean_wave.Resize(lpc_order + corr_window_size, kCopyData);
+  BaseFloat e1, e2, sum;
+  SubVector<BaseFloat> sub_vec1(zero_mean_wave, 0, corr_window_size);
+  e1 = VecVec(sub_vec1, sub_vec1);
+  (*norms)(0) += e1;
+  (*inner_prod)(0) += e1;
+  for (int32 lag = 1; lag < lpc_order; lag++) {
+    SubVector<BaseFloat> sub_vec2(zero_mean_wave, lag, corr_window_size);
+    e2 = VecVec(sub_vec2, sub_vec2);
+    sum = VecVec(sub_vec1, sub_vec2);
+    (*inner_prod)(lag) += sum;
+    (*norms)(lag) += e2;
```

Comment:
Actually I don't think you need to compute these norms at all.
From here https://en.wikipedia.org/wiki/Autocorrelation
it says "In signal processing, the above definition is often used without the normalization, that is, without subtracting the mean and dividing by the variance. When the autocorrelation function is normalized by mean and variance, it is sometimes referred to as the autocorrelation coefficient".
It looks to me from feature-plp.cc that the autocorrelation coefficients (not autocorrelations) are used as the input to the filter computation.  Generally they will be much larger than 1 because the signal is not divided by 2^16.  This means that it won't tend to hit that floor.
I am not at all happy with that floor of 1e-5 in the code but removing it will require testing.